### ISSUE TITLE
Split Linear.Epsilon into own package.

### ISSUE THREAD

**Issue by phadej** [ekmett/linear#161]
I find often needing it when doing something numeric (e.g. `ad`), but depending on `linear` feels wrong. A package (`epsilon`) would be very nice.

I'll probably will make it anyway, but it would be nice to dedup functionality on Hackage.

Another example is ApproxEq in `aeson` which does slightly different thing, but could reuse `Epsilon` class for implementations.

EDIT: also `half` could depend on it.

**Comment by RyanGlScott** [comment]
I have also wondered about making `Epsilon` a standalone piece of functionality. For what it's worse, the somewhat commonly used (and lightweight) `ieee754` library also has an [`epsilon`](https://hackage.haskell.org/package/ieee754-0.8.0/docs/Numeric-IEEE.html#v:epsilon) method in the `IEEE` class. That being said, `ieee754` and `linear` disagree on what particular values to use—for instance, `ieee754` picks `FLT_EPSILON` has the value for `Float`, whereas `linear` picks `1e-6`. Moreover, `linear`'s `Epsilon` class can work for more complicated data types, such as `Complex` and length-indexed vectors, whereas it's unclear how these would be made instances of `IEEE`.

**Comment by phadej** [comment]
Yes, so `ieee754` probably uses https://en.wikipedia.org/wiki/Machine_epsilon, but `linear`'s `nearZero` is picking larger one (probably) to be more resilient about accumulated errors etc. The `ApproxEq` usage is common, eg.

```
tests/Plucker.hs:             , "quadrance" ~: nearZero (quadranceToOrigin ln2 - 10) ~?= True
```

testing that two numbers are equal. (And I think that we should check for relative error, above it's not visible as 10 is not extremely large nor extremely small number)

I fail to find any reliable source saying "Numerically stable computations using `float`s (respectively `double`s) can be generally trusted to have N digit accuracy", I would use that for precision of `ApproxEq`. The machine epsilon is too precise for that.

**Comment by AliceRixte** [comment]
I also support the idea of splitting Linear.Epsilon in its own package. As soon as one wants to test functions on floats we need this. 

I also think it would be great to be able to adjust the epsilon value depending on the need, for instance in QuickCheck it would be great to be able to specify some epsilon to explicitely say what error is fine.

I just made a PR #184  that adds a class `MkEpsilon`  that provides a particular value for epsilon. It works nicely in conjunction with `nearZero`, IMHO.

@phadej  I couldn't find the class `ApproxEq` from  aeson. Was it removed ?

**Comment by phadej** [comment]
@AliceRixte `ApproxEq` is defined and used in `aeson`'s tests only.

**Comment by phadej** [comment]
FWIW, I don't think that `Epsilon` is a right thing to do. When comparing floating point values, we either have some concrete precision requirements (say 0.0001%) or something like https://en.wikipedia.org/wiki/Unit_in_the_last_place, FLT_EPSILON is ups(1) so makes sense only for values near `1`. 

Therefore I'm closing this, I have no good idea of what I want, but I know that's not `Epsilon` as in `linear`.
